cmd/thanos/store.go

```diff
@@ -109,17 +109,21 @@ func runStore(
 			return errors.Wrap(err, "create object storage store")
 		}
 
-		begin := time.Now()
-		level.Debug(logger).Log("msg", "initializing bucket store")
-		if err := bs.InitialSync(context.Background()); err != nil {
-			return errors.Wrap(err, "bucket store initial sync")
-		}
-		level.Debug(logger).Log("msg", "bucket store ready", "init_duration", time.Since(begin).String())
-
 		ctx, cancel := context.WithCancel(context.Background())
+		readyToServe := make(chan struct{}, 1)
+
 		g.Add(func() error {
+			begin := time.Now()
+			level.Debug(logger).Log("msg", "initializing bucket store")
+			if err := bs.InitialSync(context.Background()); err != nil {
+				return errors.Wrap(err, "bucket store initial sync")
+			}
+			level.Debug(logger).Log("msg", "bucket store readyToServe", "init_duration", time.Since(begin).String())
 			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
 
+			// Ready to start Grpc server.
+			readyToServe <- struct{}{}
```

Comment:
It's usually better to just close the channel. That way you won't have to know how many other goroutines depend on being ready (and need to send a signal for each).

If you change it, you can drop the channel buffer too.